lib/Dialect/FIRRTL/Ops.cpp

```diff
@@ -58,6 +59,34 @@ static ParseResult parseCircuitOp(OpAsmParser &parser, OperationState &result) {
   return success();
 }
 
+static LogicalResult verifyCircuitOp(CircuitOp &circuit) {
+  StringRef main = circuit.name();
+
+  // Check that the circuit must have a "main" module (top module)
+  if (main.empty()) {
+    circuit.emitOpError();
+    return failure();
+  }
+
+  StringSet<> *moduleNames = new StringSet<>();
+
+  for (auto &op : *circuit.getBody()) {
```

Comment:
The names of a module are [MLIR Symbols](https://mlir.llvm.org/docs/SymbolsAndSymbolTables/) so you can look them up directly instead of building a set manually.  Something like this should work:

`circuit.lookupSymbol(main)` which should return null if there is no match.

Reply:
Wow, this is a lot better. Built in symbol table is nice. I need to (1) read through the full MLIR docs and (2) adjust my thinking from FIRRTL Scala compiler land (no built in symbol table).

I've applied your suggestion in a later commit.